### ISSUE TITLE
makeLog instruction not reverted when contract tx out of gas

### ISSUE THREAD

**Issue by iavl** [netcloth/netcloth-chain#39]

**Comment by iavl** [comment]
do not commit state when contract tx out of gas

**Comment by iavl** [comment]
test case:

contract code:
```
6080604052604051610cdd380380610cdd8339818101604052602081101561002657600080fd5b810190808051906020019092919050505033600260006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055508060018190555050610c4e8061008f6000396000f3fe6080604052600436106100705760003560e01c8063726cab381161004e578063726cab381461029e57806372bf354b146103fd5780638da5cb5b14610428578063ddca3f431461047f57610070565b80632e1a7d4d146100755780634192c5e5146100b057806369fe0e2d14610263575b600080fd5b34801561008157600080fd5b506100ae6004803603602081101561009857600080fd5b81019080803590602001909291905050506104aa565b005b3480156100bc57600080fd5b5061021a600480360360608110156100d357600080fd5b81019080803590602001906401000000008111156100f057600080fd5b82018360208201111561010257600080fd5b8035906020019184600183028401116401000000008311171561012457600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f8201169050808301925050505050505091929192908035906020019064010000000081111561018757600080fd5b82018360208201111561019957600080fd5b803590602001918460018302840111640100000000831117156101bb57600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f820116905080830192505050505050509192919290803560ff169060200190929190505050610571565b604051808381526020018273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019250505060405180910390f35b34801561026f57600080fd5b5061029c6004803603602081101561028657600080fd5b8101908080359060200190929190505050610776565b005b6103fb600480360360608110156102b457600080fd5b81019080803590602001906401000000008111156102d157600080fd5b8201836020820111156102e357600080fd5b8035906020019184600183028401116401000000008311171561030557600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f8201169050808301925050505050505091929192908035906020019064010000000081111561036857600080fd5b82018360208201111561037a57600080fd5b8035906020019184600183028401116401000000008311171561039c57600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f820116905080830192505050505050509192919290803560ff1690602001909291905050506107db565b005b34801561040957600080fd5b50610412610bb4565b6040518082815260200191505060405180910390f35b34801561043457600080fd5b5061043d610bbc565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561048b57600080fd5b50610494610be2565b6040518082815260200191505060405180910390f35b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff16141561050557600080fd5b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff166108fc829081150290604051600060405180830381858888f1935050505015801561056d573d6000803e3d6000fd5b5050565b6000806000856040518082805190602001908083835b602083106105aa5780518252602082019150602081019050602083039250610587565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020846040518082805190602001908083835b6020831061061157805182526020820191506020810190506020830392506105ee565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008460ff1660ff168152602001908152602001600020600001546000866040518082805190602001908083835b602083106106955780518252602082019150602081019050602083039250610672565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020856040518082805190602001908083835b602083106106fc57805182526020820191506020810190506020830392506106d9565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008560ff1660ff16815260200190815260200160002060010160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1691509150935093915050565b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614156107d157600080fd5b8060018190555050565b6001543410156107ea57600080fd5b60006107f4610bb4565b9050806000856040518082805190602001908083835b6020831061082d578051825260208201915060208101905060208303925061080a565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020846040518082805190602001908083835b602083106108945780518252602082019150602081019050602083039250610871565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008460ff1660ff168152602001908152602001600020600001541015610a5b576108f0610be8565b60405180604001604052808381526020013373ffffffffffffffffffffffffffffffffffffffff168152509050806000866040518082805190602001908083835b602083106109545780518252602082019150602081019050602083039250610931565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020856040518082805190602001908083835b602083106109bb5780518252602082019150602081019050602083039250610998565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008560ff1660ff1681526020019081526020016000206000820151816000015560208201518160010160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550905050505b7f2d00c4f1918c27598c90ad291434cea2c008fc020f5729332db6a56c45613fa984848484336040518080602001806020018660ff1660ff1681526020018581526020018473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001838103835288818151815260200191508051906020019080838360005b83811015610b09578082015181840152602081019050610aee565b50505050905090810190601f168015610b365780820380516001836020036101000a031916815260200191505b50838103825287818151815260200191508051906020019080838360005b83811015610b6f578082015181840152602081019050610b54565b50505050905090810190601f168015610b9c5780820380516001836020036101000a031916815260200191505b5097505050505050505060405180910390a150505050565b600042905090565b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b60015481565b604051806040016040528060008152602001600073ffffffffffffffffffffffffffffffffffffffff168152509056fea264697066735822122004b0769b1eae0408c8f6bbc6517b2715e8435658bc1bfb80bb850c3ae55c946a64736f6c63430006000033
```

abi:
```
[
	{
		"inputs": [
			{
				"internalType": "uint256",
				"name": "_initFee",
				"type": "uint256"
			}
		],
		"stateMutability": "payable",
		"type": "constructor"
	},
	{
		"anonymous": false,
		"inputs": [
			{
				"indexed": false,
				"internalType": "string",
				"name": "fromPubkey",
				"type": "string"
			},
			{
				"indexed": false,
				"internalType": "string",
				"name": "toPubkey",
				"type": "string"
			},
			{
				"indexed": false,
				"internalType": "uint8",
				"name": "recallType",
				"type": "uint8"
			},
			{
				"indexed": false,
				"internalType": "uint256",
				"name": "timestamp",
				"type": "uint256"
			},
			{
				"indexed": false,
				"internalType": "address",
				"name": "from",
				"type": "address"
			}
		],
		"name": "Recall",
		"type": "event"
	},
	{
		"inputs": [],
		"name": "currTimeInSeconds",
		"outputs": [
			{
				"internalType": "uint256",
				"name": "",
				"type": "uint256"
			}
		],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [],
		"name": "fee",
		"outputs": [
			{
				"internalType": "uint256",
				"name": "",
				"type": "uint256"
			}
		],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [],
		"name": "owner",
		"outputs": [
			{
				"internalType": "address payable",
				"name": "",
				"type": "address"
			}
		],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [
			{
				"internalType": "string",
				"name": "fromPubkey",
				"type": "string"
			},
			{
				"internalType": "string",
				"name": "toPubkey",
				"type": "string"
			},
			{
				"internalType": "uint8",
				"name": "recallType",
				"type": "uint8"
			}
		],
		"name": "queryRecall",
		"outputs": [
			{
				"internalType": "uint256",
				"name": "",
				"type": "uint256"
			},
			{
				"internalType": "address",
				"name": "",
				"type": "address"
			}
		],
		"stateMutability": "view",
		"type": "function"
	},
	{
		"inputs": [
			{
				"internalType": "string",
				"name": "fromPubkey",
				"type": "string"
			},
			{
				"internalType": "string",
				"name": "toPubkey",
				"type": "string"
			},
			{
				"internalType": "uint8",
				"name": "recallType",
				"type": "uint8"
			}
		],
		"name": "recall",
		"outputs": [],
		"stateMutability": "payable",
		"type": "function"
	},
	{
		"inputs": [
			{
				"internalType": "uint256",
				"name": "_fee",
				"type": "uint256"
			}
		],
		"name": "setFee",
		"outputs": [],
		"stateMutability": "nonpayable",
		"type": "function"
	},
	{
		"inputs": [
			{
				"internalType": "uint256",
				"name": "amount",
				"type": "uint256"
			}
		],
		"name": "withdraw",
		"outputs": [],
		"stateMutability": "nonpayable",
		"type": "function"
	}
]
```

CLI cmd:
```
nchcli vm create --code_file=./recall_payable.bc \
--from $(nchcli keys show -a alice) \
--args="1000000000000" \
--abi_file=./recall_payable.abi \
--amount=1000000000000pnch \
--gas 823800
```

tx receipt:
```
{
  "height": "3",
  "txhash": "57717230FC58265A17FF495C77027C407B323A1FC2CEBD63AD71523FE87C549B",
  "code": 2,
  "raw_log": "[{\"msg_index\":0,\"success\":false,\"log\":\"failed to execute message; message index: 0. error: out of gas\",\"events\":null}]",
  "logs": [
    {
      "msg_index": 0,
      "success": false,
      "log": "failed to execute message; message index: 0. error: out of gas",
      "events": null
    }
  ],
  "gas_wanted": "823800",
  "gas_used": "99242",
  "tx": {
    "type": "nch/StdTx",
    "value": {
      "msg": [
        {
          "type": "nch/MsgContract",
          "value": {
            "from": "nch12dmr99v3eh39f97jnh5ga32ny2ddzznppumf2h",
            "to": "",
            "payload": "6080604052604051610cdd380380610cdd8339818101604052602081101561002657600080fd5b810190808051906020019092919050505033600260006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055508060018190555050610c4e8061008f6000396000f3fe6080604052600436106100705760003560e01c8063726cab381161004e578063726cab381461029e57806372bf354b146103fd5780638da5cb5b14610428578063ddca3f431461047f57610070565b80632e1a7d4d146100755780634192c5e5146100b057806369fe0e2d14610263575b600080fd5b34801561008157600080fd5b506100ae6004803603602081101561009857600080fd5b81019080803590602001909291905050506104aa565b005b3480156100bc57600080fd5b5061021a600480360360608110156100d357600080fd5b81019080803590602001906401000000008111156100f057600080fd5b82018360208201111561010257600080fd5b8035906020019184600183028401116401000000008311171561012457600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f8201169050808301925050505050505091929192908035906020019064010000000081111561018757600080fd5b82018360208201111561019957600080fd5b803590602001918460018302840111640100000000831117156101bb57600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f820116905080830192505050505050509192919290803560ff169060200190929190505050610571565b604051808381526020018273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019250505060405180910390f35b34801561026f57600080fd5b5061029c6004803603602081101561028657600080fd5b8101908080359060200190929190505050610776565b005b6103fb600480360360608110156102b457600080fd5b81019080803590602001906401000000008111156102d157600080fd5b8201836020820111156102e357600080fd5b8035906020019184600183028401116401000000008311171561030557600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f8201169050808301925050505050505091929192908035906020019064010000000081111561036857600080fd5b82018360208201111561037a57600080fd5b8035906020019184600183028401116401000000008311171561039c57600080fd5b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f820116905080830192505050505050509192919290803560ff1690602001909291905050506107db565b005b34801561040957600080fd5b50610412610bb4565b6040518082815260200191505060405180910390f35b34801561043457600080fd5b5061043d610bbc565b604051808273ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200191505060405180910390f35b34801561048b57600080fd5b50610494610be2565b6040518082815260200191505060405180910390f35b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff16141561050557600080fd5b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff166108fc829081150290604051600060405180830381858888f1935050505015801561056d573d6000803e3d6000fd5b5050565b6000806000856040518082805190602001908083835b602083106105aa5780518252602082019150602081019050602083039250610587565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020846040518082805190602001908083835b6020831061061157805182526020820191506020810190506020830392506105ee565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008460ff1660ff168152602001908152602001600020600001546000866040518082805190602001908083835b602083106106955780518252602082019150602081019050602083039250610672565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020856040518082805190602001908083835b602083106106fc57805182526020820191506020810190506020830392506106d9565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008560ff1660ff16815260200190815260200160002060010160009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1691509150935093915050565b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff163373ffffffffffffffffffffffffffffffffffffffff1614156107d157600080fd5b8060018190555050565b6001543410156107ea57600080fd5b60006107f4610bb4565b9050806000856040518082805190602001908083835b6020831061082d578051825260208201915060208101905060208303925061080a565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020846040518082805190602001908083835b602083106108945780518252602082019150602081019050602083039250610871565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008460ff1660ff168152602001908152602001600020600001541015610a5b576108f0610be8565b60405180604001604052808381526020013373ffffffffffffffffffffffffffffffffffffffff168152509050806000866040518082805190602001908083835b602083106109545780518252602082019150602081019050602083039250610931565b6001836020036101000a0380198251168184511680821785525050505050509050019150509081526020016040518091039020856040518082805190602001908083835b602083106109bb5780518252602082019150602081019050602083039250610998565b6001836020036101000a038019825116818451168082178552505050505050905001915050908152602001604051809103902060008560ff1660ff1681526020019081526020016000206000820151816000015560208201518160010160006101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550905050505b7f2d00c4f1918c27598c90ad291434cea2c008fc020f5729332db6a56c45613fa984848484336040518080602001806020018660ff1660ff1681526020018581526020018473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff168152602001838103835288818151815260200191508051906020019080838360005b83811015610b09578082015181840152602081019050610aee565b50505050905090810190601f168015610b365780820380516001836020036101000a031916815260200191505b50838103825287818151815260200191508051906020019080838360005b83811015610b6f578082015181840152602081019050610b54565b50505050905090810190601f168015610b9c5780820380516001836020036101000a031916815260200191505b5097505050505050505060405180910390a150505050565b600042905090565b600260009054906101000a900473ffffffffffffffffffffffffffffffffffffffff1681565b60015481565b604051806040016040528060008152602001600073ffffffffffffffffffffffffffffffffffffffff168152509056fea264697066735822122004b0769b1eae0408c8f6bbc6517b2715e8435658bc1bfb80bb850c3ae55c946a64736f6c63430006000033000000000000000000000000000000000000000000000000000000e8d4a51000",
            "amount": {
              "denom": "pnch",
              "amount": "1000000000000"
            }
          }
        }
      ],
      "fee": {
        "amount": [
          {
            "denom": "pnch",
            "amount": "823800000"
          }
        ],
        "gas": "823800"
      },
      "signatures": [
        {
          "pub_key": {
            "type": "tendermint/PubKeySecp256k1",
            "value": "ArvclYKGJIYgkp4qSpyE/gOE9QK/vBpXOPd84DGaE0TQ"
          },
          "signature": "QOspmzjo/eihbZgCnfvXIpnxRQpx4f5cZ4tOZU6bWrpivBt4wLyrdj8mus4hC8N+h3Xo/bSNaQ6mWu255c9hQA=="
        }
      ],
      "memo": ""
    }
  },
  "timestamp": "2020-05-09T09:52:24Z"
}
```

**Comment by iavl** [comment]
fixed in #41
